File: 2023/rust/src/bin/day02.rs
use aoc2023::Solve;
use std::io::{BufRead, BufReader, Read};
use std::num::ParseIntError;
use std::str::FromStr;

struct Game {
    id: u32,
    subsets: Vec<Cubes>,
}

impl FromStr for Game {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (l, r) = s.split_once(": ").unwrap();
        let id = l[5..].parse()?;
        let subsets = r.split("; ").filter_map(|s| s.parse().ok()).collect();
        Ok(Self { id, subsets })
    }
}

struct Cubes {
    red: u32,
    green: u32,
    blue: u32,
}

impl FromStr for Cubes {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (mut red, mut green, mut blue) = (0, 0, 0);
        for part in s.split(", ") {
            let (n, color) = part.split_once(' ').unwrap();
            let n = n.parse()?;
            match color {
                "red" => red = n,
                "green" => green = n,
                "blue" => blue = n,
                _ => unreachable!(),
            }
        }
        Ok(Self { red, green, blue })
    }
}

struct Solution {
    games: Vec<Game>,
}

impl Solve for Solution {
    type Answer1 = u32;
    type Answer2 = u32;

    fn new(r: impl Read) -> Self {
        let games = BufReader::new(r)
            .lines()
            .map_while(Result::ok)
            .map(|s| s.parse().unwrap())
            .collect();
        Self { games }
    }
    fn part1(&self) -> Self::Answer1 {
        self.games
            .iter()
            .filter_map(|game| {
                if game
                    .subsets
                    .iter()
                    .all(|subset| subset.red <= 12 && subset.green <= 13 && subset.blue <= 14)
                {
                    Some(game.id)
                } else {
                    None
                }
            })
            .sum()
    }
    fn part2(&self) -> Self::Answer2 {
        self.games
            .iter()
            .map(|game| {
                let (r, g, b) = game.subsets.iter().fold((0, 0, 0), |(r, g, b), cubes| {
                    (r.max(cubes.red), g.max(cubes.green), b.max(cubes.blue))
                });
                r * g * b
            })
            .sum()
    }
}

fn main() {
    let solution = Solution::new(std::io::stdin().lock());
    println!("Part 1: {}", solution.part1());
    println!("Part 2: {}", solution.part2());
}

#[cfg(test)]
mod tests {
    use super::*;

    fn example_input() -> &'static [u8] {
        r"
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
"[1..]
            .as_bytes()
    }

    #[test]
    fn part1() {
        assert_eq!(Solution::new(example_input()).part1(), 8);
    }

    #[test]
    fn part2() {
        assert_eq!(Solution::new(example_input()).part2(), 2286);
    }
}

File: 2023/ocaml/lib/day02.ml
open Base
open Solution

module Solution : Solve = struct
  module Subset = struct
    type t = { r : int; g : int; b : int }

    let create r g b = { r; g; b }
    let zero = create 0 0 0
    let max x y = create (max x.r y.r) (max x.g y.g) (max x.b y.b)
    let ( >= ) x y = x.r >= y.r && x.g >= y.g && x.b >= y.b

    let of_string s =
      let f s =
        let num, color = String.(strip s |> rsplit2_exn ~on:' ') in
        match color with
        | "red" -> create (Int.of_string num) 0 0
        | "green" -> create 0 (Int.of_string num) 0
        | "blue" -> create 0 0 (Int.of_string num)
        | _ -> failwith "invalid color"
      in
      String.split s ~on:',' |> List.map ~f |> List.fold ~init:zero ~f:max

    let power t = t.r * t.g * t.b
  end

  type t = (int * Subset.t list) list

  let parse input =
    let parse_line line =
      let game, sets = String.lsplit2_exn line ~on:':' in
      ( String.chop_prefix_exn game ~prefix:"Game " |> Int.of_string,
        String.split sets ~on:';' |> List.map ~f:Subset.of_string )
    in
    In_channel.input_lines input |> List.map ~f:parse_line

  let part1 games =
    let f (_, subsets) =
      List.for_all subsets ~f:Subset.(( >= ) (create 12 13 14))
    in
    games |> List.filter ~f |> List.sum (module Int) ~f:fst |> Answer.of_int

  let part2 games =
    let f (_, subsets) =
      List.fold subsets ~init:Subset.zero ~f:Subset.max |> Subset.power
    in
    games |> List.sum (module Int) ~f |> Answer.of_int
end
